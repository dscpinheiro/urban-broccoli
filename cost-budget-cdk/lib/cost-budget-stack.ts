import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_budgets as budgets } from 'aws-cdk-lib';

export class BudgetAppStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        // Parameters
        const thresholdAmount = new cdk.CfnParameter(this, 'BudgetThreshold', {
            type: 'Number',
            default: 50,
            minValue: 0
        });

        const notificationAddress = new cdk.CfnParameter(this, 'NotificationAddress');

        const notificationType = new cdk.CfnParameter(this, 'NotificationType', {
            type: 'String',
            default: 'EMAIL',
            allowedValues: ['EMAIL', 'SNS']
        });

        this.templateOptions.metadata = {
            'AWS::CloudFormation::Interface': {
                ParameterGroups: [
                    {
                        Label: { default: 'AWS Budget configuration' },
                        Parameters: [
                            thresholdAmount.logicalId,
                            notificationAddress.logicalId,
                            notificationType.logicalId
                        ]
                    }
                ],
                ParameterLabels: {
                    [thresholdAmount.logicalId]: {
                        default: 'Budgeted amount (in USD)'
                    },
                    [notificationAddress.logicalId]: {
                        default: 'Address that AWS sends budget notifications to, either an SNS topic or an email'
                    },
                    [notificationType.logicalId]: {
                        default: 'Type of notification that AWS sends to a subscriber'
                    }
                }
            }
        };

        // Resources
        const budgetSubscribers = [{
            address: notificationAddress.valueAsString,
            subscriptionType: notificationType.valueAsString
        }];

        const myBudget = new budgets.CfnBudget(this, 'MonthlyBudget', {
            budget: {
                budgetName: `monthly-budget-${cdk.Aws.ACCOUNT_ID}`,
                budgetType: 'COST',
                budgetLimit: {
                    amount: thresholdAmount.valueAsNumber,
                    unit: 'USD'
                },
                timeUnit: 'MONTHLY'
            },
            notificationsWithSubscribers: [
                {
                    notification: {
                        comparisonOperator: 'GREATER_THAN',
                        notificationType: 'FORECASTED',
                        threshold: 50,
                        thresholdType: 'PERCENTAGE'
                    },
                    subscribers: budgetSubscribers
                },
                {
                    notification: {
                        comparisonOperator: 'GREATER_THAN',
                        notificationType: 'FORECASTED',
                        threshold: 100,
                        thresholdType: 'PERCENTAGE'
                    },
                    subscribers: budgetSubscribers
                },

                {
                    notification: {
                        comparisonOperator: 'GREATER_THAN',
                        notificationType: 'ACTUAL',
                        threshold: 100,
                        thresholdType: 'PERCENTAGE'
                    },
                    subscribers: budgetSubscribers
                },
                {
                    notification: {
                        comparisonOperator: 'GREATER_THAN',
                        notificationType: 'ACTUAL',
                        threshold: 200,
                        thresholdType: 'PERCENTAGE'
                    },
                    subscribers: budgetSubscribers
                },
                {
                    notification: {
                        comparisonOperator: 'GREATER_THAN',
                        notificationType: 'ACTUAL',
                        threshold: 300,
                        thresholdType: 'PERCENTAGE'
                    },
                    subscribers: budgetSubscribers
                }
            ]
        });

        // Outputs
        new cdk.CfnOutput(this, 'BudgetName', { value: myBudget.ref });
    }
}
