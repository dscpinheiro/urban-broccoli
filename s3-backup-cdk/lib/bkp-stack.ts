import * as cdk from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3';
import * as kms from '@aws-cdk/aws-kms';
import * as iam from '@aws-cdk/aws-iam';

export class BackupStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        const keyAdmin = iam.User.fromUserName(this, 'BackupKeyAdmin', 'keyadmin');

        const backupKey = new kms.Key(this, 'S3BackupKey', {
            admins: [keyAdmin],
            alias: 's3-backup-key',
            description: 'Key to be used for S3 SSE-KMS',
            enabled: true
        });

        const inventoryBucket = new s3.Bucket(this, 'InventoryBucket', {
            blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
            encryption: s3.BucketEncryption.S3_MANAGED,
            lifecycleRules: [
                {
                    id: 'InventoryLifecycleRule',
                    enabled: true,
                    abortIncompleteMultipartUploadAfter: cdk.Duration.days(7),
                    expiration: cdk.Duration.days(365),
                    transitions: [
                        {
                            storageClass: s3.StorageClass.INFREQUENT_ACCESS,
                            transitionAfter: cdk.Duration.days(180)
                        }
                    ]
                }
            ]
        });

        const backupBucket = new s3.Bucket(this, 'BackupBucket', {
            blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
            bucketKeyEnabled: true,
            encryption: s3.BucketEncryption.KMS,
            encryptionKey: backupKey,
            lifecycleRules: [
                {
                    id: 'BackupLifecycleRule',
                    enabled: true,
                    abortIncompleteMultipartUploadAfter: cdk.Duration.days(7),
                    noncurrentVersionExpiration: cdk.Duration.days(90),
                    noncurrentVersionTransitions: [
                        {
                            storageClass: s3.StorageClass.ONE_ZONE_INFREQUENT_ACCESS,
                            transitionAfter: cdk.Duration.days(45)
                        }
                    ],
                    transitions: [
                        {
                            storageClass: s3.StorageClass.INFREQUENT_ACCESS,
                            transitionAfter: cdk.Duration.days(60)
                        },
                        {
                            storageClass: s3.StorageClass.GLACIER,
                            transitionAfter: cdk.Duration.days(365)
                        }
                    ]
                }
            ],
            objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_PREFERRED,
            versioned: true,
            inventories: [
                {
                    inventoryId: 'backup-inventory',
                    frequency: s3.InventoryFrequency.WEEKLY,
                    includeObjectVersions: s3.InventoryObjectVersion.CURRENT,
                    enabled: true,
                    destination: {
                        bucket: inventoryBucket,
                        bucketOwner: cdk.Aws.ACCOUNT_ID
                    },
                    optionalFields: ['Size', 'LastModifiedDate', 'StorageClass', 'ETag']
                }
            ]
        });

        new cdk.CfnOutput(this, 'KmsKeyArn', { value: backupKey.keyArn });
        new cdk.CfnOutput(this, 'BackupBucketArn', { value: backupBucket.bucketArn });
        new cdk.CfnOutput(this, 'InventoryBucketArn', { value: inventoryBucket.bucketArn });
    }
}
