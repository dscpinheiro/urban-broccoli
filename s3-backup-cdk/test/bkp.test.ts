import * as cdk from '@aws-cdk/core';
import { SynthUtils } from '@aws-cdk/assert';
import { BackupStack } from '../lib/bkp-stack';

test('creates backup bucket and kms key', () => {
    const app = new cdk.App();
    const stack = new BackupStack(app, 'TestStack');
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
