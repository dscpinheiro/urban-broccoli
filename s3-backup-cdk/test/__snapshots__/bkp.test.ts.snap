// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates backup bucket and kms key 1`] = `
Object {
  "Outputs": Object {
    "BackupBucketArn": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "BackupBucket26B8E51C",
          "Arn",
        ],
      },
    },
    "InventoryBucketArn": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "InventoryBucketA869B8CB",
          "Arn",
        ],
      },
    },
    "KmsKeyArn": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "S3BackupKeyB921581F",
          "Arn",
        ],
      },
    },
  },
  "Resources": Object {
    "BackupBucket26B8E51C": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "BucketKeyEnabled": true,
              "ServerSideEncryptionByDefault": Object {
                "KMSMasterKeyID": Object {
                  "Fn::GetAtt": Array [
                    "S3BackupKeyB921581F",
                    "Arn",
                  ],
                },
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "InventoryConfigurations": Array [
          Object {
            "Destination": Object {
              "BucketAccountId": Object {
                "Ref": "AWS::AccountId",
              },
              "BucketArn": Object {
                "Fn::GetAtt": Array [
                  "InventoryBucketA869B8CB",
                  "Arn",
                ],
              },
              "Format": "CSV",
            },
            "Enabled": true,
            "Id": "backup-inventory",
            "IncludedObjectVersions": "Current",
            "OptionalFields": Array [
              "Size",
              "LastModifiedDate",
              "StorageClass",
              "ETag",
            ],
            "ScheduleFrequency": "Weekly",
          },
        ],
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "AbortIncompleteMultipartUpload": Object {
                "DaysAfterInitiation": 7,
              },
              "Id": "BackupLifecycleRule",
              "NoncurrentVersionExpirationInDays": 60,
              "NoncurrentVersionTransitions": Array [
                Object {
                  "StorageClass": "ONEZONE_IA",
                  "TransitionInDays": 30,
                },
              ],
              "Status": "Enabled",
              "Transitions": Array [
                Object {
                  "StorageClass": "INTELLIGENT_TIERING",
                  "TransitionInDays": 30,
                },
              ],
            },
          ],
        },
        "OwnershipControls": Object {
          "Rules": Array [
            Object {
              "ObjectOwnership": "BucketOwnerPreferred",
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BackupKeyAdminPolicyC2660017": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:TagResource",
                "kms:UntagResource",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "S3BackupKeyB921581F",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BackupKeyAdminPolicyC2660017",
        "Users": Array [
          "keyadmin",
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "InventoryBucketA869B8CB": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "AbortIncompleteMultipartUpload": Object {
                "DaysAfterInitiation": 7,
              },
              "ExpirationInDays": 60,
              "Id": "InventoryLifecycleRule",
              "Status": "Enabled",
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "InventoryBucketPolicyEDF94353": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "InventoryBucketA869B8CB",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:PutObject",
              "Condition": Object {
                "ArnLike": Object {
                  "aws:SourceArn": Object {
                    "Fn::GetAtt": Array [
                      "BackupBucket26B8E51C",
                      "Arn",
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Service": "s3.amazonaws.com",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "InventoryBucketA869B8CB",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "InventoryBucketA869B8CB",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "S3BackupKeyAliasB27C0F2A": Object {
      "Properties": Object {
        "AliasName": "alias/s3-backup-key",
        "TargetKeyId": Object {
          "Fn::GetAtt": Array [
            "S3BackupKeyB921581F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
    },
    "S3BackupKeyB921581F": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "Description": "Key to be used for S3 SSE-KMS",
        "Enabled": true,
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
                "kms:GenerateDataKey",
                "kms:TagResource",
                "kms:UntagResource",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:TagResource",
                "kms:UntagResource",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":user/keyadmin",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
